keyboards/albert_smart/albert_smart84/albert_smart84.c

```diff
@@ -0,0 +1,339 @@
+        pin_t pin;
+        pin = col_pins[x];
+        setPinInputHigh(pin);
+        palEnableLineEvent(pin, PAL_EVENT_MODE_BOTH_EDGES);
+    }
+#endif
+    /* going to anabiosis*/
+    //chSysLock();
+    /* Clear Wake-up flag */
+    palEnableLineEvent(A6, PAL_EVENT_MODE_BOTH_EDGES);
+    palEnableLineEvent(B14, PAL_EVENT_MODE_BOTH_EDGES);
+    _pal_lld_enablepadevent(0,18,PAL_EVENT_MODE_BOTH_EDGES);
+    PWR->CR |= PWR_CR_CWUF | PWR_CR_CSBF;
+
+    /* Select Sleep mode */
+    //PWR->CR |= PWR_CR_PDDS | PWR_CR_LPDS; //  Can NOT wake up
+    /* Set SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+    
+    /* Request Wait For Interrupt */
+    __WFI();
+    EXTI->IMR &= ~(1<<18);
+    EXTI->EMR &= ~(1<<18);
+    EXTI->RTSR &= ~(1<<18);
+    EXTI->FTSR &= ~(1<<18);
+    EXTI->PR    = 1<<18;
+    /* Reset SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+
+    stm32_clock_init();
+    matrix_init();
+    usbWakeupHost(&USB_DRIVER);  //解决休眠无法唤醒的问题
+    restart_usb_driver(&USB_DRIVER);
+
+    setPinOutput(LED1);
+    setPinOutput(LED2);
+    setPinOutput(LED_EN);
+
+    loop10hz_token = defer_exec(LOOP_10HZ_PERIOD, loop_10Hz, NULL);
+}
+
+uint8_t sleep_flag = 0;
+
+void sleep_mode(void){
+
+    if(usb_suspend_flag == true){
+        if(timer_elapsed32(usb_suspend_time) > 500){
+            POWER_EnterSleep();
+            usb_suspend_flag = false;
+        }
+    }
+    if(USB_DRIVER.state == USB_SUSPENDED){
+        if(sleep_flag == 1){
+            usb_suspend_time = timer_read32();
+            sleep_flag = 0;
+            usb_suspend_flag = true;
+        }
+    }
+    if(USB_DRIVER.state == USB_ACTIVE){
+        // usb_suspend_flag = false;
+        sleep_flag = 1;
+    }
+    else{
+        sleep_flag = 2;
+    }
+}
+
+void keyboard_pre_init_kb(void) {
+    setPinInputLow(A9);
+    setPinInputLow(A10);
+    setPinInput(LED_EN);
+
+    setPinInputHigh(A7);
+    setPinInputHigh(A6);
+}
+
+void keyboard_post_init_kb(void) {
+
+
+    AFIO->MAPR = (AFIO->MAPR & ~AFIO_MAPR_SWJ_CFG_Msk);
+    AFIO->MAPR|= AFIO_MAPR_SWJ_CFG_DISABLE;
+
+    setPinOutput(LED1);
+    setPinOutput(LED2);
+    setPinOutput(LED_EN);
+    writePinLow(LED_EN);
+    writePinLow(LED1);
+    writePinLow(LED2);
+
+    setPinInputHigh(B14);
+
+    lock_win = dynamic_keymap_get_keycode(0,6,3);
+
+    loop10hz_token = defer_exec(LOOP_10HZ_PERIOD, loop_10Hz, NULL);
+
+}
+
+uint32_t loop_10Hz(uint32_t trigger_time, void *cb_arg){
+    sleep_mode();
+    get_mac_win();
+
+    blink_count = (blink_count+1) % (blink_period*2);
+    if(default_layer == 4)keymap_config.no_gui=false;
+
+    if(blink_count == 0){
+        if(sleep>0){
+            if(USB_DRIVER.state == USB_ACTIVE){
+                sleep--;
+                report_keyboard_t send_report;
+                memset(&send_report, 0, sizeof(report_keyboard_t));
+                host_keyboard_send(&send_report);
```

Comment:
Uhh... So you're truing to send the report constantly?  Is there an actual reason for this?

Reply:
Beacause I found out that when the MCU enters deep sleep, pressing the button to wake it up and then releasing it might have a chance of the key will lag.

Reply:
so I try to use this way to resolve it.

Reply:
> Beacause I found out that when the MCU enters deep sleep, pressing the button to wake it up and then releasing it might have a chance of the key will lag.

Make sure that you're on the latest version of qmk, there have been some changes to the usb stack for ARM boards, so that behavior may be resolved.  Additionally, those fixes are, I believe, what makes the WFI call problematic.